File: frontend/types/analysis.ts
// lowercase для БД/API, UI сам поднимет регистры где нужно
export type AnalysisSignal = "buy" | "sell" | "hold";
export type AnalysisStatus = "pending" | "success" | "failed";

export interface RiskBlock {
  entry: number;
  stop: number;
  targets: number[]; // [t1, t2, t3]
  rr: number; // risk:reward
  position_size: number; // qty
  max_loss: number; // $
  max_win: number; // $
  risk_pct: number; // % от депозита
  confidence: number; // 0..100
}

export interface TABlock {
  rsi: number;
  macd: { line: number; signal: number; hist: number };
  atr: number;
  ma: {
    ma20: number;
    ma50: number;
    ma200: number;
    bias: "bull" | "bear" | "neutral";
  };
  pattern?: string;
}

export interface DerivsBlock {
  funding_rate: number; // %
  open_interest: number; // %
  liq_nearby?: { price: number; side: "long" | "short"; size: number }[];
}

export interface OnChainBlock {
  netflow_24h: number;
  sopr?: number;
  nupl?: number;
}

export interface Analysis {
  id: number;
  symbol: string;
  interval: string;
  signal: AnalysisSignal;
  target_price: number | null;
  status: AnalysisStatus;
  created_at: string;

  // новые блоки
  risk?: RiskBlock;
  ta?: TABlock;
  derivs?: DerivsBlock;
  onchain?: OnChainBlock;
  notes?: string;
}

File: backend/app/schemas/analysis.py
# backend/app/schemas/analysis.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, Dict, Any
from enum import Enum


class AnalysisStatus(str, Enum):
    pending = "pending"
    success = "success"
    failed = "failed"


class AnalysisBase(BaseModel):
    symbol: str
    interval: str
    signal: str
    target_price: Optional[float]
    status: str
    created_at: datetime

    risk: Optional[Dict[str, Any]] = None
    ta: Optional[Dict[str, Any]] = None
    derivs: Optional[Dict[str, Any]] = None
    onchain: Optional[Dict[str, Any]] = None
    notes: Optional[str] = None


class AnalysisCreate(BaseModel):
    symbol: str
    interval: str
    deposit: float
    riskPct: float


class AnalysisDB(BaseModel):
    """
    Главная схема, которую возвращаем клиенту
    """

    id: int
    symbol: str
    interval: str
    signal: str
    target_price: Optional[float]
    status: str
    created_at: Optional[datetime]

    risk: Optional[Dict[str, Any]] = None
    ta: Optional[Dict[str, Any]] = None
    derivs: Optional[Dict[str, Any]] = None
    onchain: Optional[Dict[str, Any]] = None
    notes: Optional[str] = None

    class Config:
        from_attributes = True  # ✅ конвертация SQLAlchemy → Pydantic


# 🔹 Если нужно краткое отображение (например, в списках)
class AnalysisOut(BaseModel):
    id: int
    symbol: str
    interval: str
    signal: str
    target_price: Optional[float]
    created_at: datetime
    status: str

    class Config:
        from_attributes = True

